apollo-router/src/services/supergraph/service.rs

```diff
@@ -619,19 +620,39 @@ async fn plan_query(
         context.extensions().lock().insert::<ParsedDocument>(doc);
     }
 
-    planning
+    let qpr = planning
         .call(
             query_planner::CachingRequest::builder()
                 .query(query_str)
                 .and_operation_name(operation_name)
-                .context(context)
+                .context(context.clone())
                 .build(),
         )
         .instrument(tracing::info_span!(
             QUERY_PLANNING_SPAN_NAME,
             "otel.kind" = "INTERNAL"
         ))
-        .await
+        .await?;
+
+    // We need to operate on the BatchQuery without holding on to the extension lock, but using the
+    // batched query requires awaiting it, which would hold the lock for too long.
+    //
+    // We remove the BatchQuery here and then reinsert it after we've done our call to fix that.
+    let batching = context.extensions().lock().remove::<BatchQuery>();
```

Comment:
The model is:

[1, 2, 3] => Batch
1 => BatchQuery
2 => BatchQuery
3 => BatchQuery

(where `n` == a single query operation in a batch)

Each BatchQuery thus belongs to a single `SupergraphRequest/Response` (and is associated via `Context` extensions). Each `BatchQuery` is related to its owning `Batch` via an `Arc` shared reference.

I don't think this is racy because the only thing that *should* be interacting with that `BatchQuey` is not racing itself, but, [as I noted above](https://github.com/apollographql/router/pull/4661#discussion_r1555783780), it does seem clunky.

I will try to pursue your suggestion about Clone+Send+Sync to see if I can improve this, but I think there may be downsides to that as well.

Reply:
I've raised a PR against this PR to explore this: https://github.com/apollographql/router/pull/4941

Please let me know what you think about that.

Reply:
done: 86855eebe883f211274a51d9449eeea5e51a0aaf